### PR TITLE
ondsel: init at 2024.1.0

### LINKED ISSUE #284524: Package request: ondsel
**Project description**

Open-source engineering suite forked from FreeCAD with many improvements.

**Metadata**

* homepage URL: https://ondsel.com/
* source URL: https://github.com/Ondsel-Development/FreeCAD https://ondsel.com/download/
* license: gpl2
* platforms: unix, linux, darwin

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by pinpox**
## Description of changes

Alternative build for https://github.com/NixOS/nixpkgs/pull/284996
This PR builds ondsel from source instead of using the appimage.

I added the package next to freecad since it is a fork and shares most of its derivation.

Fixes: https://github.com/NixOS/nixpkgs/issues/284524

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by pinpox** [comment]
@ofborg build ondsel

**Comment by pinpox** [comment]
@@ofborg build

**Review by viric — APPROVED** [review]

**Comment by pinpox** [comment]
For some reason, it seems to start freecad when running the executable inside `result`, even though the correct source is pulled and build. I can't really figure out why, does anyone have ideas and could help?

**Comment by viric** [comment]
how do you expect to distinguish freecad and ondsel?

**Comment by pinpox** [comment]
> how do you expect to distinguish freecad and ondsel?

The ui looks completely different and also freecad shows a "FREECAD" as ascii art in the terminal when starting and ondsel shows ONDSEL

**Comment by Hayajiro** [comment]
After moving it to `pkgs/by-name/on/ondsel/package.nix`, Ondsel builds correctly (and that's also the correct place for the package).

As for the Ondsel branding and splash screen: That one doesn't seem to be part of the Ondsel Repo, but patched in afterwards. Whenever you build their source, you get FreeCAD + patches instead. So Ondsel pretending to be Freecad 0.22 is correct. You can confirm that you're running the correct build by checking if the Assembly workbench is available.

**Comment by viric** [comment]
ah nice I didn't know the by-name thing.

**Comment by pinpox** [comment]
> Edit: Branding stuff can be found here and needs to be applied separately: https://github.com/Ondsel-Development/freecad-feedstock

I don't know enough about conda to apply those steps to the nix package build. 
Maybe it's easier then to continue work on the PR based on the appimage since that one has everything setup?

It might be "just" a bit of branding, but the user experience and UI is completely different and defaults setup better, which is one of the main appeals of ondsel vs. freecad. 

If you have any pointers on how to apply the changes in the build of the package let me know.

I also asked @adrianinsaval over on discord, who seems to be doing the builds for the releases but haven't got any instructions on how to build ondsel from source correctly.

**Comment by b-eyselein** [comment]
The branding for Ondsel seems to be applied in the file [`recipe/build.sh`](https://github.com/Ondsel-Development/freecad-feedstock/blob/386fab80c8aaa3bb1b9454aee8695c959d642bac/recipe/build.sh#L75) form line 75 on. Maybe you could clone the repo (or just the branding folder)...

**Comment by adrianinsaval** [comment]
https://github.com/Ondsel-Development/freecad-feedstock/blob/main/recipe%2Fbuild.sh#L75-L82

This is probably what you are missing

**Comment by adrianinsaval** [comment]
You should also add the additional addons:
https://github.com/Ondsel-Development/FreeCAD-Bundle/blob/main/conda%2Flinux%2Fcreate_bundle.sh#L33-L34
https://github.com/Ondsel-Development/FreeCAD-Bundle/blob/main/conda/scripts/install_addons.py

With the additional dependencies for Ondsel-Lens:
https://github.com/Ondsel-Development/FreeCAD-Bundle/blob/458e637ab6bbceeb5fa7ab8faabb987a329ae9c4/conda/linux/create_bundle.sh#L15

**Comment by pinpox** [comment]
@b-eyselein @adrianinsaval Thansk for the pointers, that should help!

I'm also missing the `Ondsel-defaults.cfg` file from [here](https://github.com/Ondsel-Development/freecad-feedstock/blob/386fab80c8aaa3bb1b9454aee8695c959d642bac/recipe/branding/Ondsel-defaults.cfg)

I'll try to figure out where and how I'll have to place that inside the package

**Comment by pinpox** [comment]
@b-eyselein @adrianinsaval I've added the branding and defaults and now the UI looks as expected :partying_face: !

Not sure how too add the additional addons though, maybe someone can help out

**Comment by cab404** [comment]
Do you need any help with moving this forward? I think that the only diff required is to move it to by-name

**Comment by pinpox** [comment]
> Do you need any help with moving this forward? I think that the only diff required is to move it to by-name

I'm not installing any additional addons yet. Could you test/verify if anything is missing? I'll move it then when it's considered ready

**Comment by cab404** [comment]
> @b-eyselein @adrianinsaval I've added the branding and defaults and now the UI looks as expected 🥳 !
> 
> Not sure how to add the additional addons though, maybe someone can help out. Is it only the first run wizard from [here](https://github.com/Ondsel-Development/FreeCAD-Bundle/blob/458e637ab6bbceeb5fa7ab8faabb987a329ae9c4/conda/scripts/install_addons.py#L26) ?

it apparently first installs firstrun, and then installs everything under it — https://github.com/Ondsel-Development/Ondsel-FirstRun/blob/main/mods.json

File: pkgs/applications/graphics/freecad/ondsel.nix [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      substituteInPlace src/3rdParty/OndselSolver/OndselSolver.pc.in \
+        --replace-fail "\''${exec_prefix}/@CMAKE_INSTALL_LIBDIR@" "@CMAKE_INSTALL_FULL_LIBDIR@" \
+        --replace-fail "\''${prefix}/@CMAKE_INSTALL_INCLUDEDIR@" "@CMAKE_INSTALL_FULL_INCLUDEDIR@"
+    '';
+
+    postFixup =
+      let
+        feedstock = fetchFromGitHub {
+          owner = "Ondsel-Development";
+          repo = "freecad-feedstock";
```
Comment (Naxdy) [inline pkgs/applications/graphics/freecad/ondsel.nix:45]:
```suggestion
          repo = "ondsel-es-feedstock";
```

Upstream moved.

File: pkgs/applications/graphics/freecad/ondsel.nix [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+          rev = "386fab80c8aaa3bb1b9454aee8695c959d642bac";
+          hash = "sha256-H5LbcK5iXS+QFhMayvCDSfXk2ulc6BJXWVwv97nPH9o=";
+          fetchSubmodules = true;
+        };
+      in
+      ''
+        cp ${feedstock}/recipe/branding/branding.xml $out/bin
```
Comment (Naxdy) [inline pkgs/applications/graphics/freecad/ondsel.nix:52]: Is `$out/bin` correct?

Comment (pinpox) [reply pkgs/applications/graphics/freecad/ondsel.nix:52]: I think so yes, the branding is applied correctly from what I can see. Let me know if it does not for you.

**Comment by pinpox** [comment]
> it apparently first installs firstrun, and then installs everything under it — https://github.com/Ondsel-Development/Ondsel-FirstRun/blob/main/mods.json

Is there anything from this I have to add to the derivation?

File: pkgs/applications/graphics/freecad/ondsel.nix (hunk 2 above)
Comment (Naxdy) [reply pkgs/applications/graphics/freecad/ondsel.nix:52]: Just saw the `build.sh` file. Putting an `.xml` into `/bin` is totally foreign to me, but if upstream does it then I guess it's fine? ¯\_(ツ)_/¯

**Comment by ein-shved** [comment]
Hi! Are you going to continue this? Or can I create my own PR based on your?

**Comment by luzpaz** [comment]
Anyone willing to take this PR over ?

**Comment by pinpox** [comment]
Closing this, see: https://ondsel.com/blog/goodbye/

**Pull request closed without merging**